File: engine/src/main/java/de/hanno/hpengine/util/stopwatch/StopWatch.java
package de.hanno.hpengine.util.stopwatch;

import java.util.LinkedList;
import java.util.logging.Logger;

public class StopWatch {

    private static final Logger LOGGER = Logger.getLogger(StopWatch.class.getName());

	public static boolean ACTIVE = false;
	public static boolean PRINT = false;
	
	public static StopWatch stopWatch;
	static {
		stopWatch = new StopWatch();
	}
	public static StopWatch getInstance() {
		return stopWatch;
	}
	LinkedList<Watch> watches = new LinkedList<>();


	StopWatch() {}
	public void start(String description) {
		if (!ACTIVE) {return;}
		watches.addLast(new Watch( System.nanoTime(), description));
	}

	public long stop() {
		if (!ACTIVE) {return 0;}
		return (System.currentTimeMillis() - watches.pollLast().getStart());
	}

	public String stopAndGetStringMS() {
		if (!ACTIVE) {return "";}
		Watch watch = watches.pollLast();
		return String.format("%s took %.3f ms", watch.getDescripton(), (System.nanoTime() - watch.getStart())/1000000f);
	}
	
	public void stopAndPrintMS() {
		if (!ACTIVE) {return;}
		if (!PRINT) {return;}
		String out = stopAndGetStringMS();
		LOGGER.info(getIntendation() + out);
	}
	
	private String getIntendation() {
		int n = watches.size();
		StringBuilder result = new StringBuilder();
		
		while(n > 0) {
			result.append("..");
			n--;
		}
		
		return result.toString();
	}
}

File: engine/src/main/kotlin/de/hanno/hpengine/stopwatch/StopWatch.kt
package de.hanno.hpengine.stopwatch

import java.util.*
import java.util.logging.Logger

class StopWatch internal constructor() {
    companion object {
        val LOGGER = Logger.getLogger(StopWatch::class.java.name)
        var ACTIVE = false
        var PRINT = false
        var instance = StopWatch()

        var watches = LinkedList<Watch>()
        fun start(description: String) {
            if (!ACTIVE) {
                return
            }
            watches.addLast(Watch(System.nanoTime(), description))
        }

        fun stop(): Long {
            return if (!ACTIVE) {
                0
            } else System.currentTimeMillis() - watches.pollLast().start
        }

        fun stopAndGetStringMS(): String {
            if (!ACTIVE) {
                return ""
            }
            val watch = watches.pollLast()
            return String.format("%s took %.3f ms", watch.description, (System.nanoTime() - watch.start) / 1000000f)
        }

        fun stopAndPrintMS() {
            if (!ACTIVE) {
                return
            }
            if (!PRINT) {
                return
            }
            val out = stopAndGetStringMS()
            LOGGER.info(intendation + out)
        }

        private val intendation: String
            get() {
                var n = watches.size
                val result = StringBuilder()
                while (n > 0) {
                    result.append("..")
                    n--
                }
                return result.toString()
            }
    }
}